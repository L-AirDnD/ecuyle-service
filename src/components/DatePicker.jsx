import React from 'react';
import moment from 'moment';
import DateModal from './DateModal';

import {
  Wrapper,
  Title3Light,
  StyledDates,
  Paragraph,
  StyledCheckIn,
  StyledCheckOut,
  StyledArrow,
  StyledFocusText,
  StyledSelectedDateText,
} from '../styles/common';

class DatePicker extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      modalShowing: false,
      focus: '',
      checkIn: '',
      checkOut: '',
      currReservations: [],
    };

    this.handleDateClick = this.handleDateClick.bind(this);
    this.handleDayClick = this.handleDayClick.bind(this);
    this.showModal = this.showModal.bind(this);
    this.closeModal = this.closeModal.bind(this);
    this.handleClearDates = this.handleClearDates.bind(this);
  }

  componentDidMount() {
    const {
      getOpenDateModalFunc,
      getCloseDateModalFunc,
      getClearDatesFunc,
      checkIn,
      checkOut,
      reservations,
    } = this.props;
    getOpenDateModalFunc(this.showModal);
    getCloseDateModalFunc(this.closeModal);
    getClearDatesFunc(this.handleClearDates);
    this.setState({
      checkIn,
      checkOut,
      currReservations: reservations,
    });
  }

  getModalIfAppropriate() {
    const {
      modalShowing,
      checkIn,
      checkOut,
      focus,
    } = this.state;
    let { reservations } = this.props;
    if (checkIn !== '') {
      const { currReservations } = this.state;
      reservations = currReservations;
    }

    if (modalShowing) {
      return (
        <DateModal
          focus={focus}
          reservations={reservations}
          checkIn={checkIn}
          checkOut={checkOut}
          handleModalCloseClick={this.handleModalCloseClick}
          handleDayClick={this.handleDayClick}
          handleClearDates={this.handleClearDates}
        />
      );
    }
    return '';
  }

  getCheckInComponent() {
    const { focus, checkIn } = this.state;
    if (focus === 'checkIn') {
      return checkIn === ''
        ? <StyledFocusText id="checkIn" onClick={e => this.handleDateClick(e)}>Check in</StyledFocusText>
        : <StyledFocusText id="checkIn" onClick={e => this.handleDateClick(e)}>{moment(checkIn).format('MM/DD/YYYY')}</StyledFocusText>;
    }
    return checkIn === ''
      ? <Title3Light id="checkIn" onClick={e => this.handleDateClick(e)}>Check in</Title3Light>
      : <StyledSelectedDateText id="checkIn" onClick={e => this.handleDateClick(e)}>{moment(checkIn).format('MM/DD/YYYY')}</StyledSelectedDateText>;
  }

  getCheckOutComponent() {
    const { focus, checkOut } = this.state;
    if (focus === 'checkOut') {
      return checkOut === ''
        ? <StyledFocusText id="checkOut" onClick={e => this.handleDateClick(e)}>Check out</StyledFocusText>
        : <StyledFocusText id="checkOut" onClick={e => this.handleDateClick(e)}>{moment(checkOut).format('MM/DD/YYYY')}</StyledFocusText>;
    }
    return checkOut === ''
      ? <Title3Light id="checkOut" onClick={e => this.handleDateClick(e)}>Check out</Title3Light>
      : <StyledSelectedDateText id="checkOut" onClick={e => this.handleDateClick(e)}>{moment(checkOut).format('MM/DD/YYYY')}</StyledSelectedDateText>;
  }

  getClosestReservationToCheckIn(checkIn) {
    const MAX_DATE = moment(8640000000000000);
    const { reservations } = this.props;
    if (reservations.length === 1) return MAX_DATE;
    for (let i = 0; i < reservations.length - 1; i += 1) {
      const reservationStartDate = moment(reservations[i].start_date);
      const reservationEndDate = moment(reservations[i].end_date);
      const nextReservationStartDate = moment(reservations[i + 1].start_date);
      const checkInDate = moment(checkIn);
      if (checkInDate < reservationStartDate) {
        return reservationStartDate;
      }
      if (checkInDate < nextReservationStartDate && checkInDate > reservationEndDate) {
        return nextReservationStartDate;
      }
    }
    return MAX_DATE;
  }

  setCheckInDate(date) {
    const { checkOut } = this.state;
    if (checkOut !== '' && this.checkReservationDatesValidity(date, checkOut)) {
      this.setState({ checkIn: date }, () => {
        this.closeModal();
        this.sendDateDetailsToParentIfComplete();
      });
    } else {
      const currReservations = this.buildAvailableDatesAfterCheckIn(date);
      this.setState({
        checkIn: date,
        focus: 'checkOut',
        currReservations,
        checkOut: '',
      }, () => {
        this.sendDateDetailsToParentIfComplete();
      });
    }
  }

  setCheckOutDate(date) {
    const { checkIn } = this.state;
    if (checkIn === '') {
      this.setState({
        checkOut: date,
        focus: 'checkIn',
      }, () => {
        this.sendDateDetailsToParentIfComplete();
      });
    } else {
      this.setState({ checkOut: date }, () => {
        this.closeModal();
        this.sendDateDetailsToParentIfComplete();
      });
    }
  }

  checkReservationDatesValidity(checkIn, checkOut) {
    const { reservations } = this.props;
    const checkInDate = moment(checkIn);
    const checkOutDate = moment(checkOut);
    if (checkInDate > checkOutDate) return false;
    for (let i = 0; i < reservations.length; i += 1) {
      const reservationStart = moment(reservations[i].start_date);
      const reservationEnd = moment(reservations[i].end_date);
      if (checkInDate < reservationStart && checkOutDate > reservationEnd) return false;
    }
    return true;
  }

  buildAvailableDatesAfterCheckIn(checkIn) {
    const MAX_DATE = moment(8640000000000000);
    const latestCheckOut = this.getClosestReservationToCheckIn(checkIn);
    const availBeforeCheckIn = {
      start_date: moment().format(),
      end_date: moment(checkIn).subtract(1, 'days').format(),
    };
    const availAfterLatestCheckOut = {
      start_date: latestCheckOut.format(),
      end_date: MAX_DATE.format(),
    };
    return [availBeforeCheckIn, availAfterLatestCheckOut];
  }

  handleDayClick(date) {
    const { focus } = this.state;
    if (focus === 'checkIn') {
      this.setCheckInDate(date);
    } else if (focus === 'checkOut') {
      this.setCheckOutDate(date);
    }
  }

  sendDateDetailsToParentIfComplete() {
    const { checkIn, checkOut } = this.state;
    const { handleDateModalFinish } = this.props;
    handleDateModalFinish({
      checkIn,
      checkOut,
    });
  }

  handleDateClick(e) {
    const { id } = e.target;
    this.showModal(id);
  }

  handleClearDates() {
    const { reservations } = this.props;
    this.setState({
      checkIn: '',
      checkOut: '',
      currReservations: reservations,
    }, () => {
      this.closeModal();
    });
  }

  showModal(focus) {
    this.setState({
      modalShowing: true,
      focus,
    });
  }

  closeModal(focus) {
    this.setState({
      modalShowing: false,
      focus,
    });
  }

  render() {
    return (
      <Wrapper>
        <Paragraph>
          Dates
        </Paragraph>
        <StyledDates>
          <StyledCheckIn>
            { this.getCheckInComponent() }
          </StyledCheckIn>
          <StyledArrow>
            <img src="assets/right-arrow.svg" alt="right-arrow" height="20px" width="50px" />
          </StyledArrow>
          <StyledCheckOut>
            { this.getCheckOutComponent() }
          </StyledCheckOut>
        </StyledDates>
        { this.getModalIfAppropriate() }
      </Wrapper>
    );
  }
}

export default DatePicker;
