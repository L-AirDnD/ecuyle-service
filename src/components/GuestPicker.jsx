import React from 'react';

const GuestPicker = (props) => {

};

export default GuestPicker;
