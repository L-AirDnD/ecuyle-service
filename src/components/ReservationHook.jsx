import React from 'react';
import {
  Wrapper,
  TopWrapper,
  Title4,
  Title4Light,
  StyledLine,
  StyledHookLeft,
  StyledHookRight,
} from '../styles/common';

const ReservationHook = () => (
  <Wrapper>
    <StyledLine />
    <TopWrapper>
      <StyledHookLeft>
        <Title4>This home is on people&#39;s minds.</Title4>
        <Title4Light>It&#39;s been viewed 500+ times in the past week.</Title4Light>
      </StyledHookLeft>
      <StyledHookRight>
        <img src="assets/lightbulb.svg" alt="lightbulb" height="85px" width="85px" />
      </StyledHookRight>
    </TopWrapper>
  </Wrapper>
);

export default ReservationHook;
