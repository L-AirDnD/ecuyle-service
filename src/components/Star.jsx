import React from 'react';
import PropTypes from 'prop-types';

const Star = (props) => {
  let src;
  const { type } = props;
  if (type === 'full') src = 'assets/star.svg';
  if (type === 'half') src = 'assets/star-half-empty.svg';
  if (type === 'empty') src = 'assets/star-empty-2.svg';

  return (
    <img src={src} alt="star" height="10px" width="10px" />
  );
};

Star.propTypes = {
  type: PropTypes.string,
};

export default Star;
