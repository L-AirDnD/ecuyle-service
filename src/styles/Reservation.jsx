import styled from 'styled-components';
import { colors } from './common';

const StyledReservation = styled.div`
  width: 350px;
  min-width: 350px;
  border: 1px solid ${colors.border};
  display: flex;
  flex-direction: column;
  padding: 20px;
  z-index: 0;
  position: relative;
`;

export default StyledReservation;
