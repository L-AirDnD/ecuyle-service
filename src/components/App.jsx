import React from 'react';
import Reservation from './Reservation';

class App extends React.Component {
  constructor(props) {
    super(props);
    this.state = {};
  }

  render() {
    return (
      <Reservation offeringId={4} guestId={2} />
    );
  }
}

export default App;
