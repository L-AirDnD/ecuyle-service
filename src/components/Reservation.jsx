import React from 'react';
import PropTypes from 'prop-types';

import ReservationDetails from './ReservationDetails';
import DatePicker from './DatePicker';
import GuestPicker from './GuestPicker';
import ReservationConfirmation from './ReservationConfirmation';
import ReservationHook from './ReservationHook';

import StyledReservation from '../styles/Reservation';

import controller from '../controller';

class Reservation extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      offering: {},
      reservations: {},
      checkIn: '',
      checkOut: '',
      numAdults: 1,
      numChildren: 0,
      numInfants: 0,
      closeDateModalFunc: () => {},
      closeGuestModalFunc: () => {},
    };

    this.handleGuestModalClose = this.handleGuestModalClose.bind(this);
    this.handleDateModalFinish = this.handleDateModalFinish.bind(this);
    this.handleStrayClick = this.handleStrayClick.bind(this);
    this.getCloseDateModalFunc = this.getCloseDateModalFunc.bind(this);
    this.getCloseGuestModalFunc = this.getCloseGuestModalFunc.bind(this);
  }

  componentDidMount() {
    const { offeringId } = this.props;
    controller.getOfferingDetailsById(offeringId)
      .then((offering) => {
        controller.getReservationsByOfferingId(offeringId)
          .then((reservations) => {
            this.setState({
              offering,
              reservations,
            });
          })
          .catch((err) => {
            throw err;
          });
      })
      .catch((err) => {
        throw err;
      });
  }

  getCloseDateModalFunc(closeDateModalFunc) {
    this.setState({
      closeDateModalFunc,
    });
  }

  getCloseGuestModalFunc(closeGuestModalFunc) {
    this.setState({
      closeGuestModalFunc,
    });
  }

  handleGuestModalClose(guestDetails) {
    const { numAdults, numChildren, numInfants } = guestDetails;
    this.setState({
      numAdults,
      numChildren,
      numInfants,
    });
  }

  handleDateModalFinish(dateDetails) {
    const { checkIn, checkOut } = dateDetails;
    this.setState({
      checkIn,
      checkOut,
    });
  }

  handleStrayClick(e) {
    const { target } = e;
    const { id } = target;
    const { closeDateModalFunc, closeGuestModalFunc } = this.state;
    if (id !== 'checkIn' && id !== 'checkOut' && document.querySelector('#dateModal') && !document.querySelector('#dateModal').contains(target)) {
      closeDateModalFunc();
    }
    if (document.querySelector('#guestModal') && !document.querySelector('#guestModal').contains(target)) {
      closeGuestModalFunc();
    }
  }

  render() {
    console.log(this.state);
    const {
      offering: {
        pricePerDay,
        averageRating,
        totalReviewCount,
        weeklyViewCount,
        maxGuests,
        maxInfants,
      },
      numAdults,
      numChildren,
      numInfants,
      reservations,
      checkIn,
      checkOut,
    } = this.state;

    return (
      <StyledReservation onClick={this.handleStrayClick}>
        <ReservationDetails
          pricePerDay={pricePerDay}
          averageRating={averageRating}
          totalReviewCount={totalReviewCount}
        />
        <DatePicker
          checkIn={checkIn}
          checkOut={checkOut}
          reservations={reservations}
          getCloseDateModalFunc={this.getCloseDateModalFunc}
          handleDateModalFinish={this.handleDateModalFinish}
        />
        <GuestPicker
          maxGuests={maxGuests}
          maxInfants={maxInfants}
          numAdults={numAdults}
          numChildren={numChildren}
          numInfants={numInfants}
          handleGuestModalClose={this.handleGuestModalClose}
          getCloseGuestModalFunc={this.getCloseGuestModalFunc}
        />
        <ReservationConfirmation />
        { weeklyViewCount > 500
          ? <ReservationHook weeklyViewCount={weeklyViewCount} />
          : ''
        }
      </StyledReservation>
    );
  }
}

Reservation.defaultProps = {
  offeringId: 0,
};

Reservation.propTypes = {
  offeringId: PropTypes.number,
};

export default Reservation;
