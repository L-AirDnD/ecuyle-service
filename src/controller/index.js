import axios from 'axios';
import Reservation from '../components/Reservation';

const controller = {
  getOfferingDetailsById: offeringId => (new Promise((resolve, reject) => {
    axios.get(`/api/reservations/offerings/${offeringId}`)
      .then((response) => {
        resolve(response.data[0]);
      })
      .catch((err) => {
        reject(err);
      });
  })),

  getReservationsByOfferingId: offeringId => (new Promise((resolve, reject) => {
    axios.get(`/api/reservations?offering=${offeringId}`)
      .then((response) => {
        resolve(response.data);
      })
      .catch((err) => {
        reject(err);
      });
  })),

  postReservationByOfferingId: reservation => (new Promise((resolve, reject) => {
    axios.post('/api/reservations', reservation)
      .then((response) => {
        resolve(response);
      })
      .catch((err) => {
        reject(err);
      });
  })),
};

export default controller;
