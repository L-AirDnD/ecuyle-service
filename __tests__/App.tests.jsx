/* eslint-disable no-undef */
import React from 'react';
import Enzyme, { mount, shallow } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';

import moment from 'moment';

import App from '../src/components/App';
import Reservation from '../src/components/Reservation';
import DatePicker from '../src/components/DatePicker';
import GuestPicker from '../src/components/GuestPicker';
import ReservationHook from '../src/components/ReservationHook';
import Star from '../src/components/Star';
import controller from '../src/controller';

import {
  StyledConfirmation,
  Modal,
  Title3Light,
  StyledGuests,
  CircleButton,
  Title3Dark,
  LinkButton,
  StyledDay,
} from '../src/styles/common';

Enzyme.configure({ adapter: new Adapter() });

describe('App', () => {
  test('it should render the App component', () => {
    const component = shallow(<App />);
    expect(component).toMatchSnapshot();
  });
});

describe('Reservations', () => {
  let wrapper;
  beforeEach(() => {
    wrapper = mount(<Reservation offeringId={4} guestId={2} />);
  });

  test('it should render the Reservations component with appropriate number of children', () => {
    const numChildren = wrapper.find('#reservation').children().length;
    expect(numChildren).toBe(6);
  });

  test('it should open the check in modal when the Book button is selected with no dates selected', () => {
    wrapper.find(StyledConfirmation).simulate('click');
    expect(wrapper.find(Modal).length).toBe(1);
  });

  test('it should open the date modal with a focus on checkIn when "Check In" is selected', () => {
    wrapper.find(`${Title3Light}#checkIn`).simulate('click');
    expect(wrapper.find(Modal).length).toBe(1);
    expect(wrapper.find(DatePicker).state('focus')).toBe('checkIn');
    wrapper.find(`${Title3Light}#checkOut`).simulate('click');
    expect(wrapper.find(Modal).length).toBe(1);
    expect(wrapper.find(DatePicker).state('focus')).toBe('checkOut');
  });

  test('it should open the guest modal when selecting guest picker', () => {
    wrapper.find(StyledGuests).simulate('click');
    expect(wrapper.find(`${Modal}#guestModal`).length).toBe(1);
  });

  test('should not open modal when Book button is selected and both dates are picked', () => {
    wrapper.setState({
      checkIn: moment().add(2, 'days').format('YYYY-MM-DD HH:MM:SS'),
      checkOut: moment().add(4, 'days').format('YYYY-MM-DD HH:MM:SS'),
      numAdults: 2,
      numChildren: 3,
      numInfants: 4,
    }, () => {
      wrapper.find(StyledConfirmation).simulate('click');
      expect(wrapper.find(DatePicker).state('modalShowing')).toBe(false);
    });
  });

  test('it should render hook', () => {
    const hook = shallow(<ReservationHook />);
    expect(hook).toMatchSnapshot();
  });

  test('it should render a full star', () => {
    const star = mount(<Star type="full" />);
    expect(star.prop('type')).toBe('full');
    expect(star).toMatchSnapshot();
  });

  test('it should render a half star', () => {
    const star = mount(<Star type="half" />);
    expect(star.prop('type')).toBe('half');
    expect(star).toMatchSnapshot();
  });

  test('it should render a empty star', () => {
    const star = mount(<Star type="empty" />);
    expect(star.prop('type')).toBe('empty');
    expect(star).toMatchSnapshot();
  });
});

describe('Date Picker', () => {
  let wrapper;
  beforeEach(() => {
    wrapper = mount(<Reservation offeringId={4} guestId={2} />);
  });

  test('it should select checkIn date', () => {
    expect(wrapper.find(DatePicker).state('modalShowing')).toBe(false);
    wrapper.find(`${Title3Light}#checkIn`).simulate('click');
    expect(wrapper.find(Modal).length).toBe(1);
    expect(wrapper.find(DatePicker).state('focus')).toBe('checkIn');
    expect(wrapper.find(DatePicker).state('modalShowing')).toBe(true);
    wrapper.find(StyledDay).first().simulate('click');
    expect(wrapper.find(DatePicker).state('focus')).toBe('checkOut');
  });

  test('it should select checkOut date', () => {
    expect(wrapper.find(DatePicker).state('modalShowing')).toBe(false);
    wrapper.find(`${Title3Light}#checkOut`).simulate('click');
    expect(wrapper.find(Modal).length).toBe(1);
    expect(wrapper.find(DatePicker).state('focus')).toBe('checkOut');
    expect(wrapper.find(DatePicker).state('modalShowing')).toBe(true);
    wrapper.find(StyledDay).first().simulate('click');
    expect(wrapper.find(DatePicker).state('focus')).toBe('checkIn');
  });

  test('it should select clear dates button', () => {
    wrapper.find(`${Title3Light}#checkIn`).simulate('click');
    wrapper.find(DatePicker).setState({
      checkIn: moment().format('YYYY-MM-DD'),
    }, () => {
      wrapper.find(`${LinkButton}#clearDates`).simulate('click');
    });
  });
});

describe('Guest Picker', () => {
  let wrapper;
  beforeEach(() => {
    wrapper = mount(<Reservation offeringId={4} guestId={2} />);
  });

  test('it should increment Adult count when state is set', () => {
    wrapper.find(StyledGuests).simulate('click');
    expect(wrapper.find(`${Modal}#guestModal`).length).toBe(1);
    wrapper.find(`${CircleButton}#incrementAdults`).simulate('click');
    wrapper.find(GuestPicker).setState({
      numAdults: 2,
      numChildren: 0,
      numInfants: 3,
      modalShowing: true,
    }, () => {
      wrapper.find(StyledGuests).simulate('click');
      expect(wrapper.find(GuestPicker).state('numAdults')).toBe(2);
    });
  });
});

describe('Controller', () => {
  test('should successfully call getOfferingDetailsById', () => {
    controller.getOfferingDetailsById(4)
      .then((response) => {
        expect(typeof response).toBe('object');
      })
      .catch((err) => {
        throw err;
      });
  });

  test('should successfully call getReservationsByOfferingId', () => {
    controller.getReservationsByOfferingId(4)
      .then((response) => {
        expect(typeof response).toBe('object');
      })
      .catch((err) => {
        throw err;
      });
  });

  test('should fail to call getReservationsByOfferingId', () => {
    controller.getReservationsByOfferingId('hello')
      .catch((err) => {
        expect(err).toBe(!undefined);
      });
  });

  test('should successfully call postReservationByOfferingId', () => {
    const reservation = {
      offeringId: 4,
      guestId: 1,
      startDate: '2019-08-13 16:00:36.00',
      endDate: '2019-10-01 00:29:40.00',
      numAdults: 1,
      numChildren: 1,
      numInfants: 1,
      totalPrice: 400,
    };
    controller.postReservationByOfferingId(reservation)
      .then((response) => {
        expect(typeof response).toBe('object');
      })
      .catch((err) => {
        throw err;
      });
  });
});
