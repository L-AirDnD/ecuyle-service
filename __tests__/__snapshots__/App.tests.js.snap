// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App it should render the App component 1`] = `ShallowWrapper {}`;
